predicate/predicate_paging.go

```diff
@@ -0,0 +1,260 @@
+package predicate
+
+import (
+	"fmt"
+
+	"github.com/hazelcast/hazelcast-go-client/internal/hzerrors"
+	"github.com/hazelcast/hazelcast-go-client/internal/proto"
+	iserialization "github.com/hazelcast/hazelcast-go-client/internal/serialization"
+	"github.com/hazelcast/hazelcast-go-client/internal/util/nilutil"
+	"github.com/hazelcast/hazelcast-go-client/serialization"
+	"github.com/hazelcast/hazelcast-go-client/types"
+)
+
+type predPaging struct {
+	comparator        serialization.IdentifiedDataSerializable
+	internalPredicate Predicate
+	iterationType     IterationType
+	anchorList        []types.Entry
+	pageSize          int32
+	page              int32
+}
+
+// Paging creates a paging predicate with a given page size.
+func Paging(pageSize int32) *predPaging {
+	return &predPaging{
+		internalPredicate: True(),
+		comparator:        nil,
+		iterationType:     ENTRY,
+		pageSize:          pageSize,
+		page:              0,
+		anchorList:        []types.Entry{},
+	}
+}
+
+// PagingWithPredicate creates a paging predicate with a given page size and a given inner predicate.
+func PagingWithPredicate(internalPredicate Predicate, pageSize int32) *predPaging {
+	return &predPaging{
+		internalPredicate: internalPredicate,
+		comparator:        nil,
+		iterationType:     ENTRY,
+		pageSize:          pageSize,
+		page:              0,
+		anchorList:        []types.Entry{},
+	}
+}
+
+// IterationType allows the paging predicate to be used to fetch values, a keyset or an entryset.
+type IterationType string
+
+const (
+	KEY   IterationType = "KEY"
+	VALUE               = "VALUE"
+	ENTRY               = "ENTRY"
+)
+
+func iterationTypeToId(t IterationType) int8 {
+	switch t {
+	case KEY:
+		return 0
+	case VALUE:
+		return 1
+	case ENTRY:
+		return 2
+	default:
+		return -1
+	}
+}
+
+// NextPage sets the page of the paging predicate to the next page to fetch the next results in the following operation.
+func (p *predPaging) NextPage() {
+	p.page++
+}
+
+// PrevPage sets the page of the paging predicate to the previous page to fetch the previous results in the following operation.
+func (p *predPaging) PrevPage() {
+	p.page--
+}
+
+// SetPage sets the page to a given page number to fetch the results at that page in the following operation.
+func (p *predPaging) SetPage(page int32) {
+	p.page = page
+}
+
+// GetPage return the current page of the paging predicate.
+func (p predPaging) GetPage() int32 {
+	return p.page
+}
+
+// AnchorList returns the anchor list of the paging predicate containing the first item of every page fetched with the predicate.
+func (p predPaging) AnchorList() []types.Entry {
+	return p.anchorList
+}
+
+func (p predPaging) FactoryID() int32 {
+	return factoryID
+}
+
+func (p predPaging) ClassID() int32 {
+	return 15
+}
+
+func (p predPaging) ReadData(input serialization.DataInput) {
```

Comment:
It is not used because the actual type that is serialized is `PagingPredicateHolder`. The same is true in other clients, but there is also an implementation for this method (which is a bit confusing). I'll remove the implementation of the function and add a comment.